Request to endpoint /metrics fails occasionally

Requests to metrics path ```/metrics``` seems to fail from time to time and reports HTTP status 500. This occurs often when the Helidon application has been idle for a while.

## Environment Details
* Helidon Version: 1.0.0 & 2.2.0
* Helidon MP
* JDK version: 11 (build 11-28)
* OS: Linux
* Docker version: 18.09.02-CE

----------

## Problem Description
I'm using Prometheus (https://prometheus.io/) to gather metrics from various services (scrape configuration is to call ```/metrics``` endpoint twice per minute). Once or twice a day (often when the Helidon MP app has been idle for some time) the endpoint reports HTTP status 500 and an exception occurs which is caused by a NumberFormatException, see below screenshot. However, upcoming requests does work after a while.

![helidon-metrics-stacktrace](https://user-images.githubusercontent.com/4103786/53637893-bb471d80-3c24-11e9-84e0-f84363cb20ec.png)

## Steps to reproduce
Create a simple Helidon service and call ```/metrics``` on regular basic - after a few hours the endpoint reports HTTP status 500. Subsequent request will work but during the failure systems such as Prometheus will report that the service might be down which can cause negative alarms or other notifications.

## Addendum (reopening the issue)
The same general situation (long-running server -- unclear whether it has been idle or just long-running) gives the same symptom but during JSON formatting now.

A user is seeing this problem again, but in a slightly different situation. After a server has been running for some time (at least a day, it seems), attempts to retrieve the metrics via the endpoint triggers this problem as JSON attempts to format a NaN value as a `BigDecimal`. 

I am reopening this because we might need to use the same or a similar technique from another place in the code.
```
2021.02.24 10:39:23 WARNING io.helidon.webserver.RequestRouting Thread[nioEventLoopGroup-3-1,10,main]: Default error handler: Unhandled exception encountered.
java.util.concurrent.ExecutionException: Unhandled 'cause' of this exception encountered.
        at io.helidon.webserver.RequestRouting$RoutedRequest.defaultHandler(RequestRouting.java:397)
        at io.helidon.webserver.RequestRouting$RoutedRequest.nextNoCheck(RequestRouting.java:377)
        at io.helidon.webserver.RequestRouting$RoutedRequest.next(RequestRouting.java:325)
        at io.helidon.metrics.MetricsSupport$MetricsContextHandler.accept(MetricsSupport.java:619)
        at io.helidon.webserver.RequestRouting$RoutedRequest.next(RequestRouting.java:323)
        at io.helidon.webserver.cors.RequestAdapterSe.next(RequestAdapterSe.java:62)
        at io.helidon.webserver.cors.CorsSupport.prepareCORSResponseAndContinue(CorsSupport.java:79)
        at io.helidon.webserver.cors.CorsSupport.lambda$accept$0(CorsSupport.java:72)
        at java.base/java.util.Optional.ifPresentOrElse(Optional.java:203)
        at io.helidon.webserver.cors.CorsSupport.accept(CorsSupport.java:72)
        at io.helidon.webserver.RequestRouting$RoutedRequest.next(RequestRouting.java:323)
        at io.helidon.metrics.MetricsSupport.lambda$configureVendorMetrics$7(MetricsSupport.java:364)
        at io.helidon.webserver.RequestRouting$RoutedRequest.next(RequestRouting.java:323)
        at io.helidon.openapi.OpenAPISupport.registerJsonpSupport(OpenAPISupport.java:164)
        at io.helidon.webserver.RequestRouting$RoutedRequest.next(RequestRouting.java:323)
        at io.helidon.webserver.WebTracingConfig$RequestSpanHandler.accept(WebTracingConfig.java:247)
        at io.helidon.webserver.RequestRouting$RoutedRequest.next(RequestRouting.java:323)
        at io.helidon.common.context.Contexts.runInContext(Contexts.java:117)
        at io.helidon.webserver.RequestRouting.route(RequestRouting.java:89)
        at io.helidon.webserver.ForwardingHandler.channelRead0(ForwardingHandler.java:230)
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.NumberFormatException: Character N is neither a decimal digit number, decimal point, nor "e" notation exponential mark.
        at java.base/java.math.BigDecimal.<init>(BigDecimal.java:518)
        at java.base/java.math.BigDecimal.<init>(BigDecimal.java:401)
        at java.base/java.math.BigDecimal.<init>(BigDecimal.java:834)
        at java.base/java.math.BigDecimal.valueOf(BigDecimal.java:1304)
        at org.glassfish.json.JsonNumberImpl.getJsonNumber(JsonNumberImpl.java:46)
        at org.glassfish.json.JsonObjectBuilderImpl.add(JsonObjectBuilderImpl.java:104)
        at io.helidon.metrics.HelidonTimer.jsonData(HelidonTimer.java:236)
        at io.helidon.metrics.MetricsSupport.lambda$toJsonData$2(MetricsSupport.java:284)
        at java.base/java.util.stream.ReduceOps$4ReducingSink.accept(ReduceOps.java:220)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at java.base/java.util.stream.SortedOps$RefSortingSink.end(SortedOps.java:395)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:589)
        at io.helidon.metrics.MetricsSupport.toJson(MetricsSupport.java:321)
        at io.helidon.metrics.MetricsSupport.toJsonData(MetricsSupport.java:283)
        at io.helidon.metrics.MetricsSupport.accumulateJson(MetricsSupport.java:312)
        at io.helidon.metrics.MetricsSupport.lambda$toJson$5(MetricsSupport.java:305)
        at java.base/java.util.stream.ReduceOps$4ReducingSink.accept(ReduceOps.java:220)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
        at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:589)
        at io.helidon.metrics.MetricsSupport.toJson(MetricsSupport.java:304)
        at io.helidon.metrics.MetricsSupport.toJsonData(MetricsSupport.java:279)
        at io.helidon.metrics.MetricsSupport.getMultiple(MetricsSupport.java:471)
        at io.helidon.metrics.MetricsSupport.lambda$configureEndpoint$8(MetricsSupport.java:391)
        at io.helidon.webserver.RequestRouting$RoutedRequest.next(RequestRouting.java:323)
        ... 39 more
```

--- a/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018, 2020 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2021 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -185,6 +185,9 @@ class ExponentiallyDecayingReservoir {
                         final WeightedSnapshot.WeightedSample sample = values.remove(key);
                         final WeightedSnapshot.WeightedSample newSample = new WeightedSnapshot.WeightedSample(sample.getValue(),
                                                                                                               sample.getWeight() * scalingFactor);
+                        if (Double.compare(newSample.getWeight(), 0) == 0) {
+                            continue;
+                        }
                         values.put(key * scalingFactor, newSample);
                     }
                 }

--- a/metrics/metrics/src/main/java/io/helidon/metrics/HelidonTimer.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/HelidonTimer.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018, 2020 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2021 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -287,7 +287,7 @@ final class HelidonTimer extends MetricImpl implements Timer {
             this.histogram = HelidonHistogram.create(repoType, Metadata.builder()
                     .withName(name)
                     .withType(MetricType.HISTOGRAM)
-                    .build());
+                    .build(), clock);
             this.clock = clock;
         }
 

--- a/metrics/metrics/src/main/java/io/helidon/metrics/WeightedSnapshot.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/WeightedSnapshot.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018, 2020 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2021 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -65,7 +65,11 @@ class WeightedSnapshot extends Snapshot {
 
         for (int i = 0; i < copy.length; i++) {
             this.values[i] = copy[i].value;
-            this.normWeights[i] = copy[i].weight / sumWeight;
+            /*
+             * A zero denominator could cause the resulting double to be infinity or, if the numerator is also 0,
+             * NaN. Either causes problems when formatting for JSON output. Just use 0 instead.
+             */
+            this.normWeights[i] = sumWeight != 0 ? copy[i].weight / sumWeight : 0;
         }
 
         for (int i = 1; i < copy.length; i++) {

--- a/metrics/metrics/src/test/java/io/helidon/metrics/HelidonHistogramTest.java
+++ b/metrics/metrics/src/test/java/io/helidon/metrics/HelidonHistogramTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018, 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2018, 2021 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -32,6 +32,7 @@ import javax.json.Json;
 import javax.json.JsonObject;
 import javax.json.JsonObjectBuilder;
 
+import org.eclipse.microprofile.metrics.Histogram;
 import org.eclipse.microprofile.metrics.Metadata;
 import org.eclipse.microprofile.metrics.MetricID;
 import org.eclipse.microprofile.metrics.MetricType;
@@ -43,6 +44,8 @@ import org.junit.jupiter.api.Test;
 import static io.helidon.metrics.HelidonMetricsMatcher.withinTolerance;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.not;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.junit.jupiter.api.Assertions.assertAll;
 import static org.junit.jupiter.api.Assertions.fail;
@@ -225,6 +228,27 @@ class HelidonHistogramTest {
         testSnapshot(1, "delegating integers", delegatingHistoInt.getSnapshot(), 50.6, 29.4389);
         testSnapshot(10, "longs", histoLong.getSnapshot(), 506.3, 294.389);
         testSnapshot(10, "delegating longs", delegatingHistoLong.getSnapshot(), 506.3, 294.389);
+    }
+
+    @Test
+    void testNaNAvoidance() {
+        Metadata metadata = Metadata.builder()
+                .withName("long_idle_test")
+                .withDisplayName("theDisplayName")
+                .withDescription("Simulates a long idle period")
+                .withType(MetricType.HISTOGRAM)
+                .withUnit(MetricUnits.SECONDS)
+                .build();
+        TestClock testClock = TestClock.create();
+        Histogram idleHistogram = HelidonHistogram.create("application", metadata, testClock);
+
+        idleHistogram.update(100);
+
+        for (int i = 1; i < 48; i++) {
+            testClock.add(1, TimeUnit.HOURS);
+            assertThat("Idle histogram failed after simulating " + i + " hours", idleHistogram.getSnapshot().getMean(),
+                    not(equalTo(Double.NaN)));
+        }
     }
 
     private void testSnapshot(int factor, String description, Snapshot snapshot, double mean, double stddev) {

--- a/metrics/metrics/src/test/java/io/helidon/metrics/HelidonTimerTest.java
+++ b/metrics/metrics/src/test/java/io/helidon/metrics/HelidonTimerTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018, 2020 Oracle and/or its affiliates.
+ * Copyright (c) 2018, 2021 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -220,5 +220,24 @@ class HelidonTimerTest {
                                                           + "# HELP application_response_time_seconds Server response time for "
                                                           + "/index.html\n"
                                                           + "application_response_time_seconds_count 200"));
+    }
+
+    @Test
+    void testNaNAvoidance() {
+        TestClock testClock = TestClock.create();
+        HelidonTimer helidonTimer = HelidonTimer.create("application", meta, testClock);
+        MetricID metricID = new MetricID("idleTimer");
+
+        JsonObjectBuilder builder = MetricImpl.JSON.createObjectBuilder();
+        helidonTimer.update(1L, TimeUnit.MILLISECONDS);
+
+        for (int i = 1; i < 48; i++) {
+            testClock.add(1L, TimeUnit.HOURS);
+            try {
+                helidonTimer.jsonData(builder, metricID);
+            } catch (Throwable t) {
+                fail("Failed after simulating " + i + " hours");
+            }
+        }
     }
 }

